aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -206,25 +231,44 @@ module aptos_framework::fungible_asset {
         TransferRef { metadata }
     }
 
+    /// Creates a extend ref that can be used to upgrade to concurrent from the given fungible
+    /// object's constructor ref.
+    /// This can only be called at object creation time as constructor_ref is only available then.
+    public fun generate_extend_ref(constructor_ref: &ConstructorRef): ExtendRef {
+        object::generate_extend_ref(constructor_ref)
+        // let metadata = object::object_from_constructor_ref<Metadata>(constructor_ref);
+        // ExtendRef { metadata }
+    }
+
     #[view]
     /// Get the current supply from the `metadata` object.
-    public fun supply<T: key>(metadata: Object<T>): Option<u128> acquires Supply {
+    public fun supply<T: key>(metadata: Object<T>): Option<u128> acquires Supply, ConcurrentSupply {
         let metadata_address = object::object_address(&metadata);
         if (exists<Supply>(metadata_address)) {
             let supply = borrow_global<Supply>(metadata_address);
             option::some(supply.current)
+        } else if (exists<ConcurrentSupply>(metadata_address)) {
+            let supply = borrow_global<ConcurrentSupply>(metadata_address);
+            option::some(aggregator_v2::read(&supply.current))
         } else {
             option::none()
         }
     }
 
     #[view]
     /// Get the maximum supply from the `metadata` object.
-    public fun maximum<T: key>(metadata: Object<T>): Option<u128> acquires Supply {
+    public fun maximum<T: key>(metadata: Object<T>): Option<u128> acquires Supply, ConcurrentSupply {
         let metadata_address = object::object_address(&metadata);
         if (exists<Supply>(metadata_address)) {
             let supply = borrow_global<Supply>(metadata_address);
             supply.maximum
+        } else if (exists<ConcurrentSupply>(metadata_address)) {
+            let supply = borrow_global<ConcurrentSupply>(metadata_address);
+            if (supply.unlimited) {
+                option::none()
+            } else {
+                option::some(aggregator_v2::max_value(&supply.current))
```

Comment:
I wonder Can't AG itself tell whether it's unlimited or not?

Reply:
width of the field implicitly defines the max value. so internally, unlimited means max_u64.

If we want to be able to distinguish between max_u64 and unlimited asset (i.e. to show in the UI "unlimited" vs "max_u64") - we need this boolean.

but that might be better to do in the UI, not here though

Reply:
> but that might be better to do in the UI, not here though
That's what I meant... since we have an upper bound anyway so in the code the max value is always there. UI can treat max_u64 differently tho.

Reply:
It was like that,  and than I added the flag.  But we don't have a flag in token.  So for consistency- I agree, let me remove the flag